### PR TITLE
[WIP] Split shopinvader in 2 modules base and product

### PR DESCRIPTION
**Pull request by matthieusaison**
Resolves https://github.com/shopinvader/odoo-shopinvader/issues/1372

### CODE REVIEW

**Comment by lmignon** [comment]
Hi @matthieusaison 

All this seems to be a hard work. Do you plan to finalize this work in the coming days or do you need some helo on it?

Regards,

lmi

File: shopinvader_multi_category/__manifest__.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "website": "https://github.com/shopinvader/odoo-shopinvader",
     "author": "Akretion",
     "license": "AGPL-3",
-    "depends": ["shopinvader", "product_multi_category"],
+    "depends": ["shopinvader_v1_base", "product_multi_category"],
```
Comment (sebastienbeau) [inline shopinvader_multi_category/__manifest__.py]:
```suggestion
    "depends": ["shopinvader_v1_product", "product_multi_category"],
```

File: shopinvader_product_media/__manifest__.py [hunk 2]
```diff
@@ -12,6 +12,6 @@
     "license": "AGPL-3",
     "installable": True,
     "auto_install": True,
-    "depends": ["shopinvader", "storage_media_product"],
+    "depends": ["shopinvader_v1_base", "storage_media_product"],
```
Comment (sebastienbeau) [inline shopinvader_product_media/__manifest__.py]:
```suggestion
    "depends": ["shopinvader_v1_product", "storage_media_product"],
```

File: shopinvader_search_engine/__manifest__.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
     "license": "AGPL-3",
     "category": "Generic Modules",
     "depends": [
-        "shopinvader",
+        "shopinvader_v1_base",
```
Comment (sebastienbeau) [inline shopinvader_search_engine/__manifest__.py]:
```suggestion
        "shopinvader_v1_product",
```

File: shopinvader_v1_base/migrations/14.0.5.14.1/post-migrate.py [hunk 4]
```diff
@@ -9,8 +9,8 @@ def _upgrade_shopinvader_model_access(env):
     Upgrade the "shopinvader partner binding edit" access rights
```
Comment (sebastienbeau) [inline shopinvader_v1_base/migrations/14.0.5.14.1/post-migrate.py]: TODO
- not sure that the migration script should be moved, if you have the module installed, I thinks it should stay in shopinvader module

File: shopinvader_v1_base/models/shopinvader_partner.py [hunk 5]
```diff
@@ -51,7 +53,7 @@ def _is_shopinvader_user_active(self):
     _sql_constraints = [
         (
             "record_uniq",
-            "unique(backend_id, record_id, partner_email)",
+            "unique(sync_date, record_id, partner_email)",
```
Comment (sebastienbeau) [inline shopinvader_v1_base/models/shopinvader_partner.py]:
```suggestion
            "unique(backend_id, record_id, partner_email)",
```

File: shopinvader_v1_base/services/abstract_sale.py [hunk 6]
```diff
@@ -43,27 +43,34 @@ def _convert_step(self, sale):
...
-                self.env.context.get("lang"),
-                line.order_id.partner_id.lang,
-            )
-            # this likely should never happen if the request from client
-            # is forwarded properly
-            variant = line.product_id._get_invader_variant(
-                self.shopinvader_backend, line.order_id.partner_id.lang
-            )
-        product = self._convert_one_line_product(variant)
+        # variant = line.product_id._get_invader_variant(
```
Comment (sebastienbeau) [inline shopinvader_v1_base/services/abstract_sale.py]: TODO finish

File: shopinvader_v1_base/services/cart.py [hunk 7]
```diff
@@ -289,15 +289,8 @@ def _clear_cart(self, cart):
...
-    def _check_allowed_product(self, cart, params):
-        product = self.env["product.product"].browse(params["product_id"])
-        if not product._add_to_cart_allowed(
-            self.shopinvader_backend, partner=self.partner
-        ):
-            raise UserError(_("Product %s is not allowed") % product.name)
-
     def _add_item(self, cart, params):
-        self._check_allowed_product(cart, params)
+        # self._check_allowed_product(cart, params) # MOVED IN PRODUCT
```
Comment (sebastienbeau) [inline shopinvader_v1_base/services/cart.py]: TODO finish

File: shopinvader_v1_base/tests/common.py [hunk 8]
```diff
@@ -93,8 +95,8 @@ class CommonMixin(RegistryMixin, ComponentMixin, UtilsMixin):
     @staticmethod
     def _setup_backend(cls):
         cls.env = cls.env(context={"lang": "en_US"})
-        cls.backend = cls.env.ref("shopinvader.backend_1")
-        cls.backend.bind_all_product()
+        cls.backend = cls.env.ref("shopinvader_v1_base.backend_1")
+        # cls.backend.bind_all_product()
```
Comment (sebastienbeau) [inline shopinvader_v1_base/tests/common.py]: TODO remove the line

File: shopinvader_v1_product/models/shopinvader_variant.py [hunk 9]
```diff
@@ -9,8 +9,8 @@
 from odoo import api, fields, models
 from odoo.tools import float_compare, float_is_zero
 
-from ..utils import float_round
-from .tools import sanitize_attr_name
+from odoo.addons.shopinvader_v1_base.models.tools import sanitize_attr_name
```
Comment (sebastienbeau) [inline shopinvader_v1_product/models/shopinvader_variant.py]: TODO check if some tools should be moved into shopinvader_v1_product

File: shopinvader_v1_product/services/cart.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+_logger = logging.getLogger(__name__)
+
+
+class CartService(Component):
+    """Shopinvader service to provide e-commerce cart features."""
+
+    _inherit = "shopinvader.cart.service"
+
+    def _check_allowed_product(self, cart, params):
```
Comment (sebastienbeau) [inline shopinvader_v1_product/services/cart.py]: TODO call this method on add item

File: shopinvader_v1_product/tests/test_backend.py [hunk 11]
```diff
@@ -94,7 +92,7 @@ def test_bind_all_product_and_autobind_category_root_level(self):
         self.backend.category_root_binding_level = 1
         self.backend.category_binding_level = 3
         records = self._setup_bind_all_products()
-        self._bind_all_product([("id", "in", records["all_prods"].ids)])
+        # self._bind_all_product([("id", "in", records["all_prods"].ids)])
```
Comment (sebastienbeau) [inline shopinvader_v1_product/tests/test_backend.py]: TODO check

File: shopinvader_v1_product/tests/test_backend.py [hunk 12]
```diff
@@ -11,12 +11,6 @@ class BackendCase(CommonCase):
     @classmethod
     def setUpClass(cls):
         super(BackendCase, cls).setUpClass()
```
Comment (sebastienbeau) [inline shopinvader_v1_product/tests/test_backend.py]: A lot of file here import the file common, but the file is missing, maybe you forget to commit it

**Comment by matthieusaison** [comment]
Hi @lmignon 
I was in vacation last weeks. I'm back on this work.
It's a tricky work, thanks to propose your help.

**Comment by sebastienbeau** [comment]
Test are green !! \o/ 
We need to add the migration script and test it with a real project

**Comment by sebastienbeau** [comment]
@lmignon you can start to review

**Comment by lmignon** [comment]
@matthieusaison @sebastienbeau Thank you for this hard work. All this seems to go into the right way. I haven't found any issue at first read of this giant PR.

File: shopinvader_product_binding/__manifest__.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    "author": "Akretion",
+    "license": "AGPL-3",
+    "application": True,
+    "installable": True,
+    "external_dependencies": {"python": ["cerberus", "unidecode"], "bin": []},
+    "depends": [
+        "shopinvader_restapi",
+        # "component_event",  # product_event_listener
+        # "base_url",  # shopinvader.category
+        # "base_sparse_field_list_support",  # shopinvader.binding
```
Comment (hparfr) [inline shopinvader_product_binding/__manifest__.py]:
```suggestion
```

these modules are added just few lines after;

File: shopinvader_restapi/models/shopinvader_backend.py [hunk 14]
```diff
@@ -70,28 +64,6 @@ class ShopinvaderBackend(models.Model):
         "field on the sale order created.",
     )
     filter_ids = fields.Many2many(comodel_name="product.filter", string="Filter")
```
Comment (qgroulard) [inline shopinvader_restapi/models/shopinvader_backend.py]: `product.filter` is now defined in `shopinvader_product_binding`.

File: shopinvader_restapi/views/shopinvader_menu.xml [hunk 15]
```diff
@@ -22,13 +22,6 @@
         groups="sales_team.group_sale_salesman"
```
Comment (marielejeune) [inline shopinvader_restapi/views/shopinvader_menu.xml:22]: Don't forget to change web_icon path line 13 of this file (sorry but GitHub doesn't let me point out a line with no change):

`web_icon="shopinvader,static/description/icon.png"` -> `web_icon="shopinvader_restapi,static/description/icon.png"`

**Review by marielejeune — COMMENTED** [review]
Thanks for this huge work!
Just one minor comment

File: shopinvader_product_binding/__manifest__.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        "security/shopinvader_variant_security.xml",
+        "wizards/shopinvader_variant_unbinding_wizard.xml",
+        "wizards/shopinvader_variant_binding_wizard.xml",
+        "wizards/shopinvader_category_binding_wizard.xml",
+        "wizards/shopinvader_category_unbinding_wizard.xml",
+        "views/shopinvader_menu.xml",
+        "views/shopinvader_product_view.xml",
+        "views/shopinvader_variant_view.xml",
+        "views/shopinvader_category_view.xml",
+        "views/shopinvader_backend_view.xml",
```
Comment (hparfr) [inline shopinvader_product_binding/__manifest__.py:44]: why moving shopinvader backend  and shopinvader menu view ?

File: shopinvader_product_binding/__manifest__.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        "data/ir_export_category.xml",
+    ],
+    "demo": [
+        "demo/product_filter_demo.xml",
+        "demo/product_category_demo.xml",
+        "demo/product_attribute_value_demo.xml",
+        "demo/product_product_demo.xml",
+    ],
+    "pre_init_hook": "pre_init_hook",
+}
```
Comment (hparfr) [inline shopinvader_product_binding/__manifest__.py:58]: can you be more explicit about what the module shopinvader_product_binding should do and its responsibility.

**Review by hparfr — CHANGES_REQUESTED** [review]

**Comment by marielejeune** [comment]
`shopinvader.variant` and `shopinvader.variant.binding.wizard` are now defined in `shopinvader_product_binding` which is not a dependency of `shopinvader_restapi`.
This test class methods must thus be adapted: https://github.com/shopinvader/odoo-shopinvader/blob/e42e6848ed9875d848daf4c49528d2c022e38bb0/shopinvader_restapi/tests/common.py#L53C5-L74C19 and https://github.com/shopinvader/odoo-shopinvader/blob/e42e6848ed9875d848daf4c49528d2c022e38bb0/shopinvader_restapi/tests/common.py#L41C5-L51C36

File: shopinvader_product_binding/__manifest__.py (hunk 16 above)
Comment (matthieusaison) [reply shopinvader_product_binding/__manifest__.py:44]: The shopinvader menu view has been split, with only the menu linked to the product in the product_binding module.
[https://github.com/akretion/odoo-shopinvader/blob/14.0-split_shopinvader_v1_product/shopinvader_restapi/views/shopinvader_menu.xml](url)

However, the shopinvader backend view hasn't been split as it does, so I'll take care of that.

File: shopinvader_product_binding/__manifest__.py (hunk 17 above)
Comment (sebastienbeau) [reply shopinvader_product_binding/__manifest__.py:58]: maybe something like

On V14 and previous version product was bind to a shopinvader backend. On futur version this logic is remove. Product will be linked to a sale channel and the sale_channel_search_engine will take care of exporting then to the search engine.
Do not migrate this module on V16

File: shopinvader_product_binding/__manifest__.py (hunk 16 above)
Comment (hparfr) [reply shopinvader_product_binding/__manifest__.py:44]: Hi, the more I digg into it, the more I doubt it's the good solution.
~~We need one basic shoinvader module (with basic menus), shopinvader_restapi should depend of it~~

Edit: now I have a better understanding.

- Basic menus entries and security rules: shopinvader
- shopinvader_backend -> shopinvader_restapi
- ir_exports for products, categorie -> shopinvader (or shopinvader_se_product but very light module)

File: shopinvader_restapi/models/shopinvader_backend.py (hunk 14 above)
Comment (matthieusaison) [reply shopinvader_restapi/models/shopinvader_backend.py]: thanks for the feedback, [53839ec](https://github.com/shopinvader/odoo-shopinvader/pull/1382/commits/53839ec0e3327c407058187016422a6e8f7dfec1)

File: shopinvader_restapi/services/abstract_sale.py [hunk 18]
```diff
@@ -43,27 +43,14 @@ def _convert_step(self, sale):
     def _is_item(self, line):
         return True
 
+    def _get_product_information(self, line):
+        return {"id": line.product_id.id}
```
Comment (marielejeune) [inline shopinvader_restapi/services/abstract_sale.py:47]: Before, due to the shopinvader product variant, we got all info about a product inside the SO data (product name, prices, + various custom fields...). Now I see that we just get the product ID. Is it wanted? What are the benefits?

File: shopinvader_product_binding/__manifest__.py (hunk 17 above)
Comment (matthieusaison) [reply shopinvader_product_binding/__manifest__.py:58]: description added

**Comment by matthieusaison** [comment]
Yes it's wanted, the variant case is now handle
in shopinvader_product_binding, because the class ShopinvaderVariant has
moved here too.
https://github.com/shopinvader/odoo-shopinvader/blob/6d3ed9e88adb1f6e718184dba1fd37506a690de8/shopinvader_product_binding/services/abstract_sale.py#L33

Le lun. 18 sept. 2023 à 17:40, Marie Lejeune ***@***.***> a
écrit :

File: shopinvader_product_binding/__manifest__.py (hunk 16 above)
Comment (matthieusaison) [reply shopinvader_product_binding/__manifest__.py:44]: According to Seb, the menu have to stay in shopinvader_restapisecurity, 
ir_export product and category stays in shopinvader_product_binding

File: shopinvader_product_binding/__manifest__.py (hunk 13 above)
Comment (matthieusaison) [reply shopinvader_product_binding/__manifest__.py]: thanks, 
Change was done here https://github.com/shopinvader/odoo-shopinvader/pull/1382/commits/1fd8d89bc8c6e78093c63e8cbb72d04f2ba42efe

**Pull request closed without merging**